core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/SingleDirectionalParentVariableReferenceGraph.java

```diff
@@ -70,15 +78,29 @@ public SingleDirectionalParentVariableReferenceGraph(
     @Override
     public void updateChanged() {
         isUpdating = true;
+        changedEntities.sort(topologicalOrderComparator);
+        Map<Object, Object> processed = new IdentityHashMap<>();
         for (var changedEntity : changedEntities) {
```

Comment:
I wonder... if we loop this `while (changedEntity.isEmpty())` and we remove items from the end, we get to avoid the `clear()` at the end. (Of course we'd have to sort the list in reverse order.)

Reply:
(Removing items from the start would perform horribly, as the items in the array would have to shift.)